Parser/src/Parser.cpp

```diff
@@ -8,518 +8,12 @@
-		prog->declarationSegment = new DeclsSeq();
-		TokenSequence* singleDeclaration = nullptr;
-		TokenSequence* rest0 = nullptr;
-		TokenSequence* semicolon = nullptr;
-		TokenSequence* otherDeclarations = nullptr;
-		Decls* currentDecls = prog->declarationSegment;
-		do {
-			singleDeclaration = declarationPart->splitOn(ParseTree::splitIndexes->pop(), &rest0);
-			semicolon = rest0->splitOn(0, &otherDeclarations);
-			((DeclsSeq*) currentDecls)->firstDeclaration = new DeclOnly();
-			this->buildDecl(&(((DeclsSeq*)currentDecls)->firstDeclaration), singleDeclaration);
-			((DeclsSeq*) currentDecls)->restOfDeclarations = (otherDeclarations->getSize() == 0 ? ((Decls*)new DeclsSeq()) : ((Decls*)new DeclsEps()));
-			if (singleDeclaration != nullptr) {
-				singleDeclaration->prepareDelete(true);
-				delete singleDeclaration;
-			}
-			if (rest0 != nullptr) {
-				rest0->prepareDelete(true);
-				delete rest0;
-			}
-			if (semicolon != nullptr) {
-				semicolon->prepareDelete(true);
-				delete semicolon;
-			}
-			if (declarationPart != nullptr) {
-				declarationPart->prepareDelete(true);
-				delete declarationPart;
-			}
-			declarationPart = otherDeclarations;
-			currentDecls = ((DeclsSeq*) currentDecls)->restOfDeclarations;
-		} while(!currentDecls->isEps());
-	} else {
-		prog->declarationSegment = new DeclsEps();
-	}
-	if (statementPart->getSize() != 0) {
-		prog->statementSegment = new StatementsSeq();
-		TokenSequence* singleStatement = nullptr;
-		TokenSequence* rest0 = nullptr;
-		TokenSequence* semicolon = nullptr;
-		TokenSequence* otherStatements = nullptr;
-		Statements* currentStatements = prog->statementSegment;
-		do {
-			singleStatement = statementPart->splitOn(ParseTree::splitIndexes->pop(), &rest0);
-			semicolon = rest0->splitOn(0, &otherStatements);
-			this->buildStatement(&(((StatementsSeq*)currentStatements)->firstStatement), singleStatement);
-			((StatementsSeq*) currentStatements)->restOfStatements = (otherStatements->getSize() == 0 ? ((Statements*)new StatementsSeq()) : ((Statements*)new StatementsEps()));
-			if (singleStatement != nullptr) {
-				singleStatement->prepareDelete(true);
-				delete singleStatement;
-			}
-			if (rest0 != nullptr) {
-				rest0->prepareDelete(true);
-				delete rest0;
-			}
-			if (semicolon != nullptr) {
-				semicolon->prepareDelete(true);
-				delete semicolon;
-			}
-			if (statementPart != nullptr) {
-				statementPart->prepareDelete(true);
-				delete statementPart;
-			}
-			statementPart = otherStatements;
-			currentStatements = ((StatementsSeq*) currentStatements)->restOfStatements;
-		} while(!currentStatements->isEps());
-	} else {
-		prog->statementSegment = new StatementsEps();
-	}
-	if (declarationPart != nullptr) {
-		declarationPart->prepareDelete(true);
-		delete declarationPart;
-	}
-	if (statementPart != nullptr) {
-		statementPart->prepareDelete(true);
-		delete statementPart;
-	}
-	std::cout << "Parsing works\n";
-	return prog;
-}
```

Comment:
As you can see, the original concept was way overcomplicated and effectively non-debuggable.